File: src/services/tag.ts
import { invoke } from "@tauri-apps/api/core";

export interface TagInfo {
  name: string;
  target_oid: string;
  target_short_oid: string;
  is_annotated: boolean;
  tagger_name: string | null;
  tagger_date: number | null;
  message: string | null;
}

export function listTags(tabId: string): Promise<TagInfo[]> {
  return invoke<TagInfo[]>("list_tags", { tabId });
}

export function createTag(
  tabId: string,
  name: string,
  message: string | null,
): Promise<void> {
  return invoke<void>("create_tag", { tabId, name, message });
}

export function deleteTag(tabId: string, name: string): Promise<void> {
  return invoke<void>("delete_tag", { tabId, name });
}

export function checkoutTag(tabId: string, name: string): Promise<void> {
  return invoke<void>("checkout_tag", { tabId, name });
}

File: src-tauri/src/commands/tag.rs
use tauri::State;

use crate::commands::with_repo;
use crate::git::types::TagInfo;
use crate::state::AppState;

#[tauri::command]
pub fn list_tags(tab_id: String, state: State<'_, AppState>) -> Result<Vec<TagInfo>, String> {
    with_repo(&state, &tab_id, |backend| {
        backend.list_tags().map_err(|e| e.to_string())
    })
}

#[tauri::command]
pub fn create_tag(
    tab_id: String,
    name: String,
    message: Option<String>,
    state: State<'_, AppState>,
) -> Result<(), String> {
    with_repo(&state, &tab_id, |backend| {
        backend
            .create_tag(&name, message.as_deref())
            .map_err(|e| e.to_string())
    })
}

#[tauri::command]
pub fn delete_tag(tab_id: String, name: String, state: State<'_, AppState>) -> Result<(), String> {
    with_repo(&state, &tab_id, |backend| {
        backend.delete_tag(&name).map_err(|e| e.to_string())
    })
}

#[tauri::command]
pub fn checkout_tag(
    tab_id: String,
    name: String,
    state: State<'_, AppState>,
) -> Result<(), String> {
    with_repo(&state, &tab_id, |backend| {
        backend.checkout_tag(&name).map_err(|e| e.to_string())
    })
}
